### PR TITLE
refactor(mla): redesign batch MLA backend internals

### PR DESCRIPTION
**Pull request by saltyminty**
## 📌 Description

This PR reorganizes Batch MLA around isolated backend implementations and
provides consistent wrapper and functional APIs across the supported dense MLA
backends.

### Highlights

- **Isolated backend implementations:** FA2, FA3, CUTLASS, TRTLLM-GEN, XQA,
  CuTe DSL monolithic, and CuTe DSL modular now live in backend-specific modules
  that own their support checks, planning/preparation, state, and launch logic.

- **Expanded wrapper API:** `BatchMLAPagedAttentionWrapper` now supports
  `fa2`, `fa3`, `cutlass`, `trtllm-gen`, `xqa`,
  `cute-dsl-monolithic`, and `cute-dsl-modular`.
  `cute-dsl` remains a family selector between its two implementations.

- **General wrapper auto fallback:** `backend="auto"` ranks every wrapper
  backend and tries candidates until one reports that it supports the request.
  Fallback occurs only for the typed unsupported-backend result; invalid input
  and unexpected failures remain visible. The current ordering is primarily
  architecture-based. Workload- and performance-informed ranking based on the
  complete problem configuration is deferred, but the new structure makes it
  possible without modifying every backend.

- **Unified functional API:** `batch_mla_paged_attention` is the supported
  one-shot functional entrypoint for explicit FA2, FA3, CUTLASS, TRTLLM-GEN,
  XQA, and CuTe DSL execution. Functional `auto` retains its existing
  TRTLLM-GEN/CuTe autotuning policy and remains separate from wrapper `auto`.

- **Canonical planning and layouts:** `MLAPlanMetadata` provides canonical
  CSR/dense planning metadata, while `query_layout` and `kv_cache_layout`
  declare packed or split runtime representations. Every wrapper backend accepts CSR, dense, or equivalent
  dual metadata; the selected backend's native representation is derived
  lazily. Packed query and KV inputs are lowered into zero-copy views when
  possible.

- **Compatibility and deprecations:** Existing public imports remain available,
  while the following legacy forms are deprecated:
  - direct TRTLLM-GEN and XQA functional entrypoints, in favor of
    `batch_mla_paged_attention`;
  - positional wrapper arguments and flat planning metadata, in favor of
    keyword arguments and `MLAPlanMetadata`;
  - legacy split `q_nope`/`q_pe` and `ckv_cache`/`kpe_cache` arguments,
    whether positional or keyword, in favor of structural `query` and
    `kv_cache` arguments;
  - explicit CUTLASS execution without `plan()`; and
  - CUTLASS's allocating fallback for independently stored split KV tensors.

The architecture and conventions for future attention backend refactors are
documented in the repository's
[Batch MLA backend architecture](https://github.com/flashinfer-ai/flashinfer/blob/34525d173297cf2e6e7910d1cc8498dd00299124/docs/design_docs/batch_mla_backend_architecture.md).

### Downstream integrations

The public API is exercised by draft integrations in two production serving
stacks:

- [vLLM #52990](https://github.com/vllm-project/vllm/pull/52990) migrates dense
  MLA decode to `batch_mla_paged_attention` while preserving backend selection,
  return-LSE behavior, and the persistent multi-CTA counter buffer.
- [SGLang #35555](https://github.com/sgl-project/sglang/pull/35555) migrates
  planning to `MLAPlanMetadata`, uses the unified tensor run interface, and
  preserves sync-free CUDA graph planning across capture and replay.

## 🔍 Related Issues

https://github.com/flashinfer-ai/flashinfer/issues/4037

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Targeted validation completed:

- SM90 canonical wrapper/functional sweeps:
  19 passed, 38 skipped.
- SM100 canonical wrapper/functional sweeps:
  29 passed, 28 skipped.
- SM120 canonical wrapper/functional sweeps:
  15 passed, 42 skipped.
- Latest-main focused compatibility suite:
  148 passed, 1 deselected.
- H100 FA3 CUDA graph numerical validation:
  1 passed.
- B200 SGLang CUDA graph capture/replay validation:
  2 passed, including 1 parametrized subtest.
- H100 ABBA comparison against the pre-change all-GPU path found no measurable
  regression: plan median average +0.51% and run median -3.67%.
- Additional SM100 checks passed for variable-Q CuTe routing, TRTLLM-GEN
  capacity/head shapes, CuTe DCP, and native no-RoPE sparse TRTLLM-GEN.

Current validation status:

- The GitHub `pre-commit --all-files` job passes Ruff and formatting, but
  currently reports two mypy errors in
  `flashinfer/mla/_batch_mla/_auto_policy.py`.
- Public API and documentation CI passes.
- The repository GPU matrix is skipped pending authorization. A contributor who
  can label the PR, or a `ci-users` member, can comment `@flashinfer-bot run` to
  start it.
- Full repository tests and wheel/build validation have not been run for the
  final rebased tree. The targeted H100 ABBA comparison above found no
  measurable planning or execution regression in the unchanged all-GPU path.

## Reviewer Notes

Please focus review on:

- the wrapper-versus-functional lifecycle boundary;
- backend-owned planned implementations and functional runners;
- wrapper auto policy versus the intentionally narrower functional auto policy;
- CuTe DSL family dispatch versus strict concrete identifiers;
- metadata compatibility and zero-copy packed/split layout rules;
- typed unsupported fallback and transactional plan publication;
- preservation of legacy functional facades, exports, and trace identities;
- the consolidated production-test ownership model; and
- the latest-main variable-Q, TRTLLM-GEN, sparse no-RoPE, and CuTe DCP
  compatibility ports.

## Summary by CodeRabbit

* **New Features**
  * Added functional and stateful MLA paged-attention APIs with automatic backend selection across GPU architectures.
  * Added support for packed and split cache layouts, LSE output, sparse MLA, quantization, CUDA graphs, and autotuning.
  * Added backend compatibility checks, structured fallback handling, and expanded validation.
* **Documentation**
  * Expanded MLA API and benchmarking guidance, including lifecycle, CLI controls, backend capabilities, and output behavior.
* **Deprecations**
  * Marked legacy XQA and TRT-LLM MLA entry points as deprecated with migration guidance.

### CODE REVIEW

File: flashinfer/mla/_batch_mla/_core.py [hunk 1]
```diff
@@ -0,0 +1,2655 @@
...
+        num_heads: int,
+        head_dim_ckv: int,
+        head_dim_kpe: int,
+        page_size: int,
+        causal: bool,
+        sm_scale: float,
+        q_data_type: torch.dtype,
+        kv_data_type: torch.dtype,
+        use_profiler: bool = False,
+        *,
```
Comment (aleozlx) [inline flashinfer/mla/_batch_mla/_core.py]: there was a suggestion about common vs backend specific params boudnary. tbd later

@saltyminty pls file an issue

File: flashinfer/_backend.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Private contracts shared by backend-dispatched FlashInfer APIs.
+"""
+
+
+class _BackendPlanUnsupportedError(ValueError):
```
Comment (aleozlx) [inline flashinfer/_backend.py:14]: does user catch the specific class or err message is enough

**Review by aleozlx — COMMENTED** [review]
left a few comments

File: flashinfer/_backend.py (hunk 2 above)
Comment (saltyminty) [reply flashinfer/_backend.py:14]: It will be caught by any user catching ValueError

File: flashinfer/mla/__init__.py [hunk 3]
```diff
@@ -12,4 +12,38 @@
...
+    MLAPlanMetadata,
+    deepseek_mla_dimensions,
+    nope_mla_dimensions,
+    smaller_mla_dimensions,
+    supported_mla_head_dimensions,
+    supported_mla_layer_dimensions,
+    trtllm_batch_decode_sparse_mla_dsv4,
+)
+from ._batch_mla._functional import (
+    batch_mla_paged_attention,
```
[the remark this answers is not in the document]
Comment (saltyminty) [reply flashinfer/mla/__init__.py:32]: Doesn't apply

File: benchmarks/mla/reference.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+            batch_lse = denominator_lse
+            if contract.lse_mode == "base2":
+                batch_lse = batch_lse / torch.log(
+                    torch.tensor(2.0, device=logits.device)
+                )
+            lse[q_start:q_end] = batch_lse
+        probabilities = torch.exp(logits - denominator_lse.unsqueeze(-1))
+        output[q_start:q_end] = (
+            torch.einsum("qhk,kd->qhd", probabilities, logical_ckv) * bmm2
+        )
```
[the remark this answers is not in the document]
Comment (saltyminty) [reply benchmarks/mla/reference.py:308]: Doesn't apply

**Comment by aleozlx** [comment]
## PR Review Screening

**CI verdict:** ✅ auto-run ok
**Review category:** `live` (rule fired: C1.2 net-new public interface; independently C3.2 durable — shared dispatch/abstraction redesign)
**Blocking checks:** S3 — "pre-commit run --all-files" unchecked (body admits 2 mypy errors in `flashinfer/mla/_batch_mla/_auto_policy.py`); "All tests are passing" unchecked
**Release blocker:** no
**Early stop:** no

### Security
| Q | Answer | Evidence |
|---|--------|----------|
| S1 injection/supply-chain | no | — |
| S2 template overwritten | no | — |
| S3 template obligations | unmet | Pre-commit + all-tests boxes unchecked; body admits 2 mypy errors in `_auto_policy.py` |
| S4 agent-directed text | no | — |

### Packaging
| Q | Answer | Evidence |
|---|--------|----------|
| C1.1 dependency bump | no | no supply/ci bucket files in 64-file diff |
| C1.2 public API changes | yes — net-new | see list below |
| C1.3 AOT/trace registration | partial gap | no `gen_*_module()` added, `aot.py` untouched (ok); but `flashinfer/trace/templates/attention.py` +168 lines and `tests/trace/example.py` rewritten with **zero** `tests/trace/fi_trace_out/` JSONs regenerated in the diff |

C1.2 detail (net-new interface, broad scope — not the narrow-scope boundary case):
- **net-new**: `batch_mla_paged_attention(*, query=None, q_nope=None, q_pe=None, kv_cache=None, ckv_cache=None, kpe_cache=None, workspace_buffer, qk_nope_head_dim, kv_lora_rank, qk_rope_head_dim, block_tables, seq_lens, max_seq_len, sparse_mla_top_k=0, out=None, bmm1_scale=1.0, bmm2_scale=1.0, sinks=None, skip_softmax_threshold_scale_factor=None, enable_pdl=None, backend="auto", is_var_seq=True, uses_shared_paged_kv_idx=True, lse=None, return_lse=False, cute_dsl_impl="auto", kv_scale_format="auto", …)` — `flashinfer/mla/_batch_mla/_functional.py:845`, `@flashinfer_api`-decorated, exported in `flashinfer/mla/__init__.py` `__all__`
- **net-new shape + extension**: `BatchMLAPagedAttentionWrapper.plan()` gains a keyword-only `metadata: MLAPlanMetadata` overload with `query_layout`/`kv_cache_layout`/`lse_mode`/`scale_mode`/`output_dtype`/`skip_softmax` (`_wrapper.py:510`); legacy CSR positional signature retained as overload
- **net-new exports**: `MLAPlanMetadata`, `MLAAutoSelectionTrace` (absent at merge-base)
- **extension**: wrapper `backend=` accepts new identifiers `cute-dsl-monolithic`/`cute-dsl-modular`; new wrapper-level `auto` fallback ranking
- **internal-move**: `trtllm_batch_decode_with_kv_cache_mla`, `xqa_batch_decode_with_kv_cache_mla` relocated to `_batch_mla/_functional.py` (facades preserved per body)

### Presentation
| Q | Answer | Evidence |
|---|--------|----------|
| C2.1 perf claim backed | n/a | refactor PR; body makes no perf claim (explicitly says benchmarks not run) |

### Implementation
| Q | Answer | Evidence |
|---|--------|----------|
| C3.1 experimental-track | no | linked issue is design issue #4037, no experimental-terms declaration |
| C3.2 durable areas | yes — durable | shared autotuner (`flashinfer/autotuner/autotuner.py` +83/−24), `flashinfer/_backend.py`, `trace/template.py`, `trace_apply/plan_capture.py`; whole-backend dispatch convention redesign; kernel header `include/flashinfer/attention/cutlass_mla.cuh` drops `fmha.initialize(...)` call |
| C3.3 tests match change | yes | 22 test files incl. new `test_mla_dispatch.py`, `test_mla_functional.py`, `test_mla_wrapper.py`, `test_autotuner_mla_decode.py` |
| C3.4 release-blocker fix | no | refactor; no hang/crash/regression claim |

Notes for the maintainer:
- The `cutlass_mla.cuh` removal of `fmha.initialize()` is a behavioral kernel-launcher change inside a "refactor" PR — worth a pointed reviewer question.
- 19k added LOC with a design doc (`docs/design_docs/batch_mla_backend_architecture.md`) — design issue #4037 is the natural venue for the live discussion.

<sub>Generated by flashinfer-pr-screen · rubric: docs/code_review_guidance.md · not a code review · AI screening can make mistakes — a maintainer's judgment supersedes this report.</sub>

**Comment by aleozlx** [comment]
reviewed the design in a prior meeting. no new questions
